test/blackbox-tests/test-cases/pkg/lockfile-generation.t/run.t

```diff
@@ -119,3 +149,103 @@ Run the solver again. This time it will fail.
         bar.0.0.1: Incompatible with restriction: >= 0.2
   [1]
 
+We'll also test how the lockfile generation works with alternate solutions.
+`bar-or-baz` is a package that depends on either `bar` or `baz` and the solver
+should pick one of them.
+
+  $ cat >dune-project <<EOF
+  > (lang dune 3.8)
+  > (package
+  >  (name lockfile_generation_test)
+  >  (depends bar-or-baz))
+  > EOF
+  $ mkpkg bar-or-baz <<EOF
+  > opam-version: "2.0"
+  > depends: [ "bar" | "baz" ]
+  > EOF
+
+After running this we expact a solution that has either `bar` or `baz` but not
+both.
+
+  $ dune pkg lock --opam-repository-path=mock-opam-repository
+  Solution for dune.lock:
+  bar.0.5.0
```

Comment:
Is it always preferred in order?

Reply:
At least in my testing, this it has been 100% reproducible. Which is great, because we definitely want to be able to go from an exact input of opam files to have the exact same solution for reproducibility and even just easier debugging.

Reply:
Can we guarantee that this won't change in the future? I'm thinking that a `List.rev` might cause problems if added without thought by the opam devs. If it is not part of the opam spec should we let them know about it?

Reply:
I think this is more of a theoretical problem than a practical one.

The OPAM devs don't have much to do with it as the solver is provided by 0install-solver and that is vendored, so it isn't gonna change under out feet. If the output changes in a future version of the solver, the then test needs to be adjusted (and possibly a note in a changelog added about the behavior changing) or a bug report needs to be filed if this change happened by accident.

I don't know if OPAM guarantees that in a `a | b` situation it will prefer `a` over `b` unless otherwise possible - I don't think so as it would probably make it harder to swap external solvers (cc @kit-ty-kate maybe you know more). I would expect that the only guarantee is it will return *some* valid solution if there is one.

But given that the test uses a very specific solver and it is not swappable, it seems to me that the thing it checks for is pretty accurate and is not going to cause flaky test runs.

Reply:
Thanks for the explanation, I agree that this isn't really a practical problem.